File: library/src/main/java/com/afollestad/aesthetic/views/AestheticToolbar.java
package com.afollestad.aesthetic.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorInt;
import android.support.annotation.Nullable;
import android.support.annotation.RestrictTo;
import android.support.v7.widget.Toolbar;
import android.util.AttributeSet;

import com.afollestad.aesthetic.Aesthetic;

import rx.Observable;
import rx.Subscription;
import rx.subjects.PublishSubject;

import static android.support.annotation.RestrictTo.Scope.LIBRARY_GROUP;
import static com.afollestad.aesthetic.Rx.distinctToMainThread;
import static com.afollestad.aesthetic.Rx.onErrorLogAndRethrow;
import static com.afollestad.aesthetic.TintHelper.createTintedDrawable;
import static com.afollestad.aesthetic.Util.setOverflowButtonColor;

/** @author Aidan Follestad (afollestad) */
@RestrictTo(LIBRARY_GROUP)
public class AestheticToolbar extends Toolbar {

  private int backgroundResId;
  private BgIconColorState lastState;
  private Subscription subscription;
  private PublishSubject<Integer> onColorUpdated;

  public AestheticToolbar(Context context) {
    super(context);
    init(context, null);
  }

  public AestheticToolbar(Context context, @Nullable AttributeSet attrs) {
    super(context, attrs);
    init(context, attrs);
  }

  public AestheticToolbar(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
    init(context, attrs);
  }

  private void init(Context context, AttributeSet attrs) {
    onColorUpdated = PublishSubject.create();
    if (context == null) {
      return;
    }
    if (attrs != null) {
      int[] attrsArray = new int[] {android.R.attr.background};
      TypedArray ta = context.obtainStyledAttributes(attrs, attrsArray);
      backgroundResId = ta.getResourceId(0, 0);
      ta.recycle();
    }
  }

  private void invalidateColors(BgIconColorState state) {
    lastState = state;
    setBackgroundColor(state.bgColor());
    setTitleTextColor(state.iconTitleColor().activeColor());
    setOverflowButtonColor(this, state.iconTitleColor().activeColor());
    if (getNavigationIcon() != null) {
      setNavigationIcon(getNavigationIcon());
    }
    onColorUpdated.onNext(state.bgColor());
    ViewUtil.tintToolbarMenu(this, getMenu(), state.iconTitleColor());
  }

  public Observable<Integer> colorUpdated() {
    return onColorUpdated.asObservable();
  }

  @Override
  public void setNavigationIcon(@Nullable Drawable icon) {
    if (lastState == null) {
      super.setNavigationIcon(icon);
      return;
    }
    super.setNavigationIcon(createTintedDrawable(icon, lastState.iconTitleColor().toEnabledSl()));
  }

  public void setNavigationIcon(@Nullable Drawable icon, @ColorInt int color) {
    if (lastState == null) {
      super.setNavigationIcon(icon);
      return;
    }
    super.setNavigationIcon(createTintedDrawable(icon, color));
  }

  @SuppressWarnings("ConstantConditions")
  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();
    subscription =
        Observable.combineLatest(
                ViewUtil.getObservableForResId(
                    getContext(), backgroundResId, Aesthetic.get().colorPrimary()),
                Aesthetic.get().colorIconTitle(),
                BgIconColorState::create)
            .compose(distinctToMainThread())
            .subscribe(this::invalidateColors, onErrorLogAndRethrow());
  }

  @Override
  protected void onDetachedFromWindow() {
    lastState = null;
    onColorUpdated = null;
    subscription.unsubscribe();
    super.onDetachedFromWindow();
  }
}

File: library/src/main/java/com/afollestad/aesthetic/views/AestheticToolbar.kt
/**
 * Designed and developed by Aidan Follestad (@afollestad)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.afollestad.aesthetic.views

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import androidx.annotation.ColorInt
import androidx.appcompat.widget.Toolbar
import com.afollestad.aesthetic.Aesthetic.Companion.get
import com.afollestad.aesthetic.R
import com.afollestad.aesthetic.internal.AttrWizard
import com.afollestad.aesthetic.utils.darkenColor
import com.afollestad.aesthetic.utils.observableForAttrName
import com.afollestad.aesthetic.utils.setOverflowButtonColor
import com.afollestad.aesthetic.utils.subscribeTo
import com.afollestad.aesthetic.utils.tint
import com.afollestad.aesthetic.utils.tintMenu
import com.afollestad.aesthetic.utils.unsubscribeOnDetach
import io.reactivex.subjects.PublishSubject

/** @author Aidan Follestad (afollestad) */
class AestheticToolbar(
  context: Context,
  attrs: AttributeSet? = null
) : Toolbar(context, attrs) {

  private var onColorUpdated = PublishSubject.create<Int>()
  private var menuIconColor: Int? = null

  private val wizard = AttrWizard(context, attrs)
  private val backgroundColorValue = wizard.getRawValue(android.R.attr.background)
  private val titleTextColorValue = wizard.getRawValue(R.attr.titleTextColor)
  private val subtitleTextColorValue = wizard.getRawValue(R.attr.subtitleTextColor)

  fun colorUpdated() = onColorUpdated

  override fun setNavigationIcon(icon: Drawable?) {
    if (menuIconColor == null) {
      super.setNavigationIcon(icon)
      return
    }
    super.setNavigationIcon(icon.tint(menuIconColor!!))
  }

  fun setNavigationIcon(icon: Drawable?, @ColorInt color: Int) {
    if (menuIconColor == null) {
      super.setNavigationIcon(icon)
      return
    }
    super.setNavigationIcon(icon.tint(color))
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()

    get().observableForAttrName(
        backgroundColorValue,
        get().colorPrimary()
    )!!
        .distinctUntilChanged()
        .doOnNext { onColorUpdated.onNext(it) }
        .subscribeTo(::setBackgroundColor)
        .unsubscribeOnDetach(this)

    get().toolbarIconColor()
        .distinctUntilChanged()
        .subscribeTo(::invalidateColors)
        .unsubscribeOnDetach(this)

    get().observableForAttrName(
        titleTextColorValue,
        get().toolbarTitleColor()
    )
        ?.distinctUntilChanged()
        ?.subscribeTo(::setTitleTextColor)
        ?.unsubscribeOnDetach(this)

    get().observableForAttrName(
        subtitleTextColorValue,
        get().toolbarSubtitleColor()
    )
        ?.distinctUntilChanged()
        ?.subscribeTo(::setSubtitleTextColor)
        ?.unsubscribeOnDetach(this)
  }

  private fun invalidateColors(color: Int) {
    this.menuIconColor = color
    setOverflowButtonColor(color)
    tintMenu(menu, color, color.darkenColor())
    if (navigationIcon != null) {
      this.navigationIcon = navigationIcon
    }
  }
}
